packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -96,64 +115,122 @@ export function memo<T, U, V, W, X, Y, Z>(
+
+      debuggingContext = (cache as any)[DEBUG_CACHE_CONTEXT];
+
+      cache = debugCache;
+    }
+
+    let { tag, snapshot, fn } = cache;
 
-  let memoized = (...args: unknown[]): T => {
-    if (!tag || !validateTag(tag, snapshot)) {
+    if (tag === undefined || !validateTag(tag, snapshot)) {
       beginTrackFrame();
 
       try {
         if (DEBUG) {
-          runInAutotrackingTransaction!(() => (lastValue = callback(...args)), debuggingContext);
+          runInAutotrackingTransaction!(() => (cache.lastValue = fn(...args)), debuggingContext!);
         } else {
-          lastValue = callback(...args);
+          cache.lastValue = fn(...args);
         }
       } finally {
         tag = endTrackFrame();
-        snapshot = valueForTag(tag);
+        cache.tag = tag;
+        cache.snapshot = valueForTag(tag);
         consumeTag(tag);
-
-        // If the final tag is constant, then we know for sure that this
-        // memoized function can never change. There are times when this
-        // information is useful externally (i.e. in the append VM, it tells us
-        // whether or not to emit opcodes) so we expose it via a metadata weakmap.
-        if (tag === CONSTANT_TAG) {
-          IS_CONST_MAP.set(memoized, true);
-        } else if (DEBUG) {
-          // In DEBUG, set the value to false explicitly. This way we can throw
-          // if someone attempts to call `isConst(memoized)` before running
-          // `memoized()` at least once.
-          IS_CONST_MAP.set(memoized, false);
-        }
       }
     } else {
       consumeTag(tag);
     }
 
-    return lastValue!;
-  };
-
-  if (DEBUG) {
-    IS_CONST_MAP.set(memoized, undefined);
+    return cache.lastValue!;
   }
 
-  return memoized;
-}
+  static isConst(cache: Cache) {
+    if (DEBUG) {
+      let debugCache = (cache as any)[DEBUG_CACHE];
+
+      if (!(debugCache instanceof Cache)) {
+        throw new Error(
+          `isConst() can only be used on an instance of a cache created with createCache(). Called with: ${String(
+            cache
+          )}`
+        );
+      }
 
-export function isConstMemo(memoized: Function) {
-  if (DEBUG && IS_CONST_MAP.has(memoized) && IS_CONST_MAP.get(memoized) === undefined) {
-    throw new Error(
-      'Attempted to call `isConstMemo` on a memoized function, but the function has not been run at least once yet. You cannot know if a memoized function is constant or not until it has been run at least once. Call the function, then pass it to `isConstMemo`.'
-    );
+      if (debugCache.tag === undefined) {
+        throw new Error(
+          `isConst() can only be used on a cache once getValue() has been called at least once. Called with cache function:\n\n${String(
+            debugCache.fn
+          )}`
+        );
+      }
+
+      cache = debugCache;
+    }
+
+    return isConstTag(cache.tag!);
   }
+}
 
-  return IS_CONST_MAP.get(memoized) === true;
+export function createCache<T>(fn: () => T): Cache<T> {
```

Comment:
the exported function should use a branded interface for Cache and not leak the class Cache